pkg/operator/controller/externaldns/credentials-request.go

```
@@ -0,0 +1,174 @@
+	}
+
+	return true, current, nil
+}
+
+func awsProviderConfig(codec *cco.ProviderCodec) (*runtime.RawExtension, error) {
+	awsProvSpec, err := codec.EncodeProviderSpec(
+		&cco.AWSProviderSpec{
+			TypeMeta: metav1.TypeMeta{
+				Kind: "AWSProviderSpec",
+			},
+			StatementEntries: []cco.StatementEntry{
+				{
+					Effect: "Allow",
+					Action: []string{
+						"elasticloadbalancing:DescribeLoadBalancers",
+						"route53:ListHostedZones",
+						"route53:ChangeResourceRecordSets",
+						"tag:GetResources"},
+					Resource: "*"}}})
+	return awsProvSpec, err
+}
+
+func azureProviderConfig(codec *cco.ProviderCodec) (*runtime.RawExtension, error) {
+	awsProvSpec, err := codec.EncodeProviderSpec(
+		&cco.AzureProviderSpec{
+			TypeMeta: metav1.TypeMeta{
+				Kind: "AzureProviderSpec",
+			},
+			RoleBindings: []cco.RoleBinding{
+				{Role: "Contributor"},
+			},
+		})
+	return awsProvSpec, err
+}
+
+func gcpProviderConfig(codec *cco.ProviderCodec) (*runtime.RawExtension, error) {
+	awsProvSpec, err := codec.EncodeProviderSpec(
+		&cco.GCPProviderSpec{
+			TypeMeta: metav1.TypeMeta{
+				Kind: "GCPProviderSpec",
+			},
+			PredefinedRoles: []string{
+				"roles/dns.admin",
+			},
+		})
+	return awsProvSpec, err
+}
+
+// desiredCredentialsRequestName returns the desired credentials request definition for externalDNS
+func desiredCredentialsRequestName(name types.NamespacedName, externalDNS *operatorv1alpha1.ExternalDNS) *cco.CredentialsRequest {
+	credentialsRequest := &cco.CredentialsRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+		Spec: cco.CredentialsRequestSpec{
+			ServiceAccountNames: []string{"external-dns-operator"},
+			SecretRef: k8sv1.ObjectReference{
+				Name:      "externaldns-cloud-credentials",
+				Namespace: "external-dns-operator",
+			},
+		},
+	}
+
+	codec, err := cco.NewCodec()
+	if err != nil {
+		fmt.Printf("error creating new codec: %v", err)
```

1. You should not use `fmt` for logging, use the logger given by the controller runtime, it's in `reconciler` object.
2. Do you need to log here? I think you can simply return an error with descriptive message.